core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,171 @@
+            writerAssignments[i] = new IntArrayList();
+        }
+
+        // Initialize partitionAssignments and partitionWriterIndex with the partition limit
+        partitionWriterIndex = new int[partitionCount];
+        partitionAssignments = new IntArrayList[partitionCount];
+        for (int i = 0; i < partitionAssignments.length; i++) {
+            partitionAssignments[i] = new IntArrayList();
+        }
+    }
+
+    @Override
+    public void accept(Page page)
+    {
+        Page partitionPage = partitionedPagePreparer.apply(page);
+
+        // Assign each row to a partition which limits to the partitionCount. If there are more physical partitions than
+        // this artificial partition limit, then it is possible that multiple partitions will get assigned the same
+        // bucket id. Thus, multiple partitions will be scaled together since we track partition physicalWrittenBytes
+        // using the artificial limit (partitionCount). The assignments lists are all expected to cleared by the
+        // previous iterations.
+        for (int position = 0; position < partitionPage.getPositionCount(); position++) {
+            int partition = partitionFunction.getPartition(partitionPage, position);
+            partitionAssignments[partition].add(position);
+        }
+
+        // Scale up writers when current buffer memory utilization is more than 50% of the maximum
+        if (memoryManager.getBufferedBytes() > maxBufferedBytes * 0.5) {
+            partitionRebalancer.rebalancePartitions();
+        }
+
+        ImmutableMap.Builder<WriterPartitionId, Integer> writerPartitionIdToRowCount = ImmutableMap.builder();
+
+        // assign each partition to a specific writer
+        for (int partitionId = 0; partitionId < partitionAssignments.length; partitionId++) {
+            IntArrayList positionsList = partitionAssignments[partitionId];
+            int partitionSize = positionsList.size();
+            if (partitionSize == 0) {
+                continue;
+            }
+
+            // clear the assigned positions list size for the next iteration to start empty. This
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            int nextPartitionWriterId = getNextWriterId(partitionId);
+            writerPartitionIdToRowCount.put(new WriterPartitionId(nextPartitionWriterId, partitionId), partitionSize);
+
+            for (int i = 0; i < partitionSize; i++) {
+                writerAssignments[nextPartitionWriterId].add(positions[i]);
+            }
+        }
+
+        // Update the partitions row count state which will help with scaling partitions across writers
+        partitionRebalancer.addPartitionsRowCount(writerPartitionIdToRowCount.buildOrThrow());
```

Comment:
`MAX_PARTITIONS_TO_REBALANCE_PER_WRITER * 32 = 4096`, which means in worst case `writerPartitionIdToRowCount` will be as big as page.

Consider reversing the relationship, e.g: `partitionRebalancer` fetching `writerPartitionIdToRowCount` from `ScaleWriterPartitioningExchanger` instead.

Reply:
I thought about this but it just makes the code more complex. I wonder if it's worth the effort.

Reply:
I reversed the relationship